tests/csapi/device_lists_test.go

```diff
@@ -6,61 +6,315 @@ import (
+		bob := deployment.RegisterUser(t, otherHSName, generateLocalpart("bob"), "password", false)
+		checkBobKeys := uploadNewKeys(t, bob)
+
+		roomID := bob.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+
+		// Alice performs an initial sync
+		_, aliceNextBatch := alice.MustSync(t, client.SyncReq{})
+
+		// Alice joins the room
+		alice.JoinRoom(t, roomID, []string{otherHSName})
+		bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(alice.UserID, roomID))
+
+		// Check that Alice receives a device list update from Bob
+		alice.MustSyncUntil(
+			t,
+			client.SyncReq{Since: aliceNextBatch},
+			syncDeviceListsHas("changed", bob.UserID),
+		)
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+		// Some homeservers (Synapse) may emit another `changed` update after querying keys.
+		_, aliceNextBatch = alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+		// Both homeservers think Alice has joined now
+		// Bob then updates their device list
+		checkBobKeys = uploadNewKeys(t, bob)
+
+		// Check that Alice receives a device list update from Bob
+		alice.MustSyncUntil(
+			t,
+			client.SyncReq{Since: aliceNextBatch},
+			syncDeviceListsHas("changed", bob.UserID),
+		)
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+	}
+
+	// testOtherUserLeave tests another user leaving a room Alice is in.
+	testOtherUserLeave := func(t *testing.T, deployment *docker.Deployment, hsName string, otherHSName string) {
+		alice := deployment.RegisterUser(t, hsName, generateLocalpart("alice"), "password", false)
+		bob := deployment.RegisterUser(t, otherHSName, generateLocalpart("bob"), "password", false)
+		checkBobKeys := uploadNewKeys(t, bob)
+
+		roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+
+		// Bob joins the room
+		bob.JoinRoom(t, roomID, []string{hsName})
+		bobNextBatch := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+		// Alice performs an initial sync
+		alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+		// Some homeservers (Synapse) may emit another `changed` update after querying keys.
+		_, aliceNextBatch := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+		// Bob leaves the room
+		bob.LeaveRoom(t, roomID)
+		bob.MustSyncUntil(t, client.SyncReq{Since: bobNextBatch}, client.SyncLeftFrom(bob.UserID, roomID))
+
+		// Check that Alice is notified
+		alice.MustSyncUntil(
+			t,
+			client.SyncReq{Since: aliceNextBatch},
+			syncDeviceListsHas("left", bob.UserID),
+		)
+
+		// Both homeservers think Bob has left now
+		// Bob then updates their device list
+		// Alice's homeserver is not expected to get the device list update and must not return a
+		// cached device list for Bob.
+		checkBobKeys = uploadNewKeys(t, bob)
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
```

Comment:
To check: this is testing that

- Alice sees Bob's new keys
- Alice doesn't see Bob's old keys

but it doesn't check that Alice's HS didn't receive a device list update?

Reply:
Let's add a check for a lack of a device list update on the next sync.